pkg/provisioner/fss/fss.go

```diff
@@ -28,248 +28,388 @@ import (
+		CompartmentId:      common.String(fsp.client.CompartmentOCID()),
+		AvailabilityDomain: &ad,
+		DisplayName:        &displayName,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	switch count := len(resp.Items); {
+	case count == 1:
+		return &resp.Items[0], nil
+	case count > 1:
+		return nil, errors.Errorf("found more than one file system with display name %q", displayName)
+	}
+
+	return nil, errNotFound
+}
+
+func (fsp *filesystemProvisioner) getOrCreateFileSystem(ctx context.Context, logger *zap.SugaredLogger, ad, displayName string) (*fss.FileSystem, error) {
+	fs, err := fsp.getFileSystemByDisplayName(ctx, ad, displayName)
+	if err != nil && err != errNotFound {
+		return nil, err
+	}
+	if fs != nil {
+		return fsp.awaitFileSystem(ctx, logger, *fs.Id)
+	}
+
+	resp, err := fsp.client.FSS().CreateFileSystem(ctx, fss.CreateFileSystemRequest{
+		CreateFileSystemDetails: fss.CreateFileSystemDetails{
 			CompartmentId:      common.String(fsp.client.CompartmentOCID()),
-			Page:               page,
+			AvailabilityDomain: &ad,
+			DisplayName:        &displayName,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	logger.With("fileSystemID", *resp.FileSystem.Id).Info("Created FileSystem")
+
+	return fsp.awaitFileSystem(ctx, logger, *resp.FileSystem.Id)
+}
+
+// findExport looks for an existing export with the same filesystem ID, export set ID, and path.
+// NOTE: No two non-'DELETED' export resources in the same export set can reference the same file system.
+func (fsp *filesystemProvisioner) findExport(ctx context.Context, fsID, exportSetID string) (*fss.ExportSummary, error) {
+	var page *string
+	for {
+		resp, err := fsp.client.FSS().ListExports(ctx, fss.ListExportsRequest{
+			CompartmentId: common.String(fsp.client.CompartmentOCID()),
+			FileSystemId:  &fsID,
+			ExportSetId:   &exportSetID,
+			Page:          page,
 		})
 		if err != nil {
 			return nil, err
 		}
-		mountTargets = append(mountTargets, resp.Items...)
+		for _, export := range resp.Items {
+			if export.LifecycleState == fss.ExportSummaryLifecycleStateCreating ||
+				export.LifecycleState == fss.ExportSummaryLifecycleStateActive {
+				return &export, nil
+			}
+		}
 		if page = resp.OpcNextPage; resp.OpcNextPage == nil {
 			break
 		}
 	}
-	return mountTargets, nil
+
+	return nil, errNotFound
 }
 
-func (fsp *filesystemProvisioner) getOrCreateMountTarget(ctx context.Context, mtID string, ad string, subnetID string) (*filestorage.MountTarget, error) {
-	if mtID != "" {
-		// Mount target already specified in the configuration file, find it in the list of mount targets
-		return fsp.getMountTargetFromID(ctx, mtID)
-	}
-	mountTargets, err := fsp.listAllMountTargets(ctx, ad)
-	if err != nil {
+func (fsp *filesystemProvisioner) getOrCreateExport(ctx context.Context, logger *zap.SugaredLogger, fsID, exportSetID string) (*fss.Export, error) {
+	export, err := fsp.findExport(ctx, fsID, exportSetID)
+	if err != nil && err != errNotFound {
```

Comment:
Do we need the second conjunction here? I.e can we simply do:

```go
if err != nil {
    return nil, err
}
```

Reply:
Yes as otherwise we’d never create an Export if one doesn’t exist.